### DISCUSSION TITLE
How can I solve source.on is not a function when using form data?

### DISCUSSION THREAD

**Discussion by positivedeveloper36** [adonisjs/core#3842] [Help]
My code like this :
```

const axios = require('axios').default
import FormData from 'form-data'
import fs from 'fs'

export default class MyController {
  public async handleMultipleFile({ request }: HttpContextContract) {
    if(request.files('files').length > 0){
      const formData = new FormData()
      
      for( let i = 0; i < request.files('files').length; i++ ) {
        let file = request.files('files')[i]
        formData.append('files[' + i + ']', file)
      }

      axios.post(`https://example-api.com/assets/files`, formData, {
          headers: {
              'Content-Type': 'multipart/form-data'
          },
        }
      ).then(res => console.log(res.data))
      .catch(err => console.error(err))
    }
  }
}

```

When I call the code from postman, I find this error ```"message": "source.on is not a function"```

I add ```return``` to see the formData in the postman

I had install library form data and axios in my code and I had import it

I want to post multiple file to a external api

How can I solve this problem? Please help. Thanks

Additional description 

So I'm building an api to handle multiple files . Later these files will be stored in a file server which will be sent via my api. So in my api there is a third party api to store the files

The full error  ::

```
{
    "message": "source.on is not a function",
    "stack": "TypeError: source.on is not a function\n    at Function.DelayedStream.create (/Users/secret/myapp-api/node_modules/delayed-stream/lib/delayed_stream.js:33:10)\n    at FormData.CombinedStream.append (/Users/secret/myapp-api/node_modules/combined-stream/lib/combined_stream.js:45:37)\n    at FormData.append (/Users/secret/myapp-api/node_modules/form-data/lib/form_data.js:75:3)\n    at MyController.upload (/Users/secret/myapp-api/app/Controllers/Http/MyController.ts:35:18)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at Object.PreCompiler.runRouteHandler [as fn] (/Users/secret/myapp-api/node_modules/@adonisjs/http-server/build/src/Server/PreCompiler/index.js:44:31)\n    at Server.handleRequest (/Users/secret/myapp-api/node_modules/@adonisjs/http-server/build/src/Server/index.js:108:13)"
}
```

**Comment by Barbapapazes** [comment]
Hello, can you provide more details about this code or a reproduction repository? Thanks

**Reply by positivedeveloper36** [reply]
@Barbapapazes I had updated my question

**Reply by Barbapapazes** [reply]
Thanks, so you're AdonisJS server is a proxy to upload files in another API which are unaccessible directly to client? (to be sure I understand everything)

Can you provide more detail about your error, like a stack trace? Because I don't understand where this `source.on` come from!

**Reply by positivedeveloper36** [reply]
@Barbapapazes Yes. I had updated my question

**Reply by Barbapapazes** [reply]
Okay thanks and have you tried the second explanation in the usage section of the docs?

https://www.npmjs.com/package/form-data

**Reply by positivedeveloper36** [reply]
@Barbapapazes Do you mean this : https://github.com/form-data/form-data#readme?

**Reply by Barbapapazes** [reply]
Yes that's what I mean!

**Reply by positivedeveloper36** [reply]
@Barbapapazes So I must use the createReadStream method? right?

**Reply by Barbapapazes** [reply]
I think the http response stream could be a better way!

**Reply by positivedeveloper36** [reply]
@Barbapapazes Yes. But my problem is I don't use path. The image uploaded from the front end is not stored in my local folder, but sent directly to the external api. So it seems I have to convert the uploaded file to blob/binary first so that it can be submitted using form data. Right?

**Reply by Barbapapazes** [reply]
I think that write e driver for [direct upload](https://docs.adonisjs.com/guides/direct-file-uploads) could be a better solution!

**Comment by RomainLanz** [comment]
Hey @positivedeveloper36! 👋🏻 

Not sure where AdonisJS fits in your question.

This place is dedicated to questions you may have for AdonisJS. It is not meant to be used to ask generic questions on JavaScript.

**Reply by positivedeveloper36** [reply]
I tried in vue js and javascript, it worked. But in adonis, this doesn't work. I'm looking a reference about the form data in adonis, but haven't found it yet. That's why I'm asking here

**Reply by thetutlage** [reply]
Vue is a frontend framework. AdonisJS is a backend framework. So I am not sure if the code written for the browser can work on the server.

The docs on Request body and bodyparser are here (if that is what you are looking for) https://docs.adonisjs.com/guides/request#request-body
